### PR TITLE
Update ZTSI module

### PR DESCRIPTION
**Pull request by robertschaedler3**
## Description

Refactor ZTSI module to apply patterns from sample module.

## Checklist

- [x] I have read the [contribution guidelines](https://github.com/Azure/azure-osconfig/blob/main/CONTRIBUTING.md).
- [x] I added unit-tests to validate my changes. All unit tests are passing.
- [x] I have merged the latest `main` branch prior to this PR submission.
- [x] I submitted this PR against the `main` branch.

### CODE REVIEW

File: src/modules/ztsi/src/lib/Ztsi.cpp [hunk 1]
```diff
@@ -105,6 +111,63 @@ bool IsValidClientName(const std::string& clientName)
...
+            }
+        }
+        catch (const std::exception& e)
+        {
+            OsConfigLogError(ZtsiLog::Get(), "Could not allocate payload: %s", e.what());
+            status = EINTR;
+
+            if (nullptr != *payload)
+            {
+                delete[] * payload;
```
Comment (MariusNi) [inline src/modules/ztsi/src/lib/Ztsi.cpp:151]: >delete[] * payload [](http://example.com/codeflow?start=16&length=18)

nit: no space between '*' and 'payload' ('*payload' is the pointer being freed)

File: src/modules/ztsi/src/lib/Ztsi.cpp [hunk 2]
```diff
@@ -116,6 +179,198 @@ Ztsi::Ztsi(std::string filePath, unsigned int maxPayloadSizeBytes)
...
+        "UserAccount": 0})"""";
+
+    if (nullptr == clientName)
+    {
+        OsConfigLogError(ZtsiLog::Get(), "GetInfo called with null clientName");
+        status = EINVAL;
+    }
+    else if (!IsValidClientName(clientName))
+    {
+        status = EINVAL;
```
Comment (MariusNi) [inline src/modules/ztsi/src/lib/Ztsi.cpp:204]: >status = EINVAL; [](http://example.com/codeflow?start=8&length=16)

Add an error log for this

File: src/modules/ztsi/src/so/ZtsiModule.cpp [hunk 3]
```diff
@@ -231,105 +162,11 @@ int MmiSet(
...
-        {
-            delete[] * payload;
-            *payload = nullptr;
-        }
-
-        if (nullptr != payloadSizeBytes)
-        {
-            *payloadSizeBytes = 0;
-        }
+        session = reinterpret_cast<Ztsi*>(clientSession);
```
Comment (MariusNi) [inline src/modules/ztsi/src/so/ZtsiModule.cpp:168]: >session = reinterpret_cast<Ztsi*>(clientSession); [](http://example.com/codeflow?start=8&length=49)

What if some client calls MpiSet with a non null but invalid handle? This cast would prodice an invalid module object. Add a "IsValid()" member to the module  object to validate itself? (by size, contents, etc)

Comment (MariusNi) [reply src/modules/ztsi/src/so/ZtsiModule.cpp:168]: Same for MpiGet

File: src/modules/ztsi/tests/ZtsiTests.cpp [hunk 4]
```diff
@@ -75,6 +81,34 @@ namespace OSConfig::Platform::Tests
...
     std::string ZtsiTests::filename = "./ztsi/config.temp.json";
 
+    TEST_F(ZtsiTests, GetSetServiceUrl)
+    {
+        char serviceUrl[] = "\"https://localhost:8080/\"";
+        MMI_JSON_STRING payload = nullptr;
+        int payloadSizeBytes = 0;
+
+        ASSERT_EQ(MMI_OK, ztsi->Set(g_componentName, g_desiredServiceUrl, serviceUrl, sizeof(serviceUrl)));
+        ASSERT_EQ(MMI_OK, ztsi->Get(g_componentName, g_reportedServiceUrl, &payload, &payloadSizeBytes));
```
Comment (MariusNi) [inline src/modules/ztsi/tests/ZtsiTests.cpp:91]: Add Set and Get tests that use invalid arguments? Null values, unsupported values, etc.

Comment (MariusNi) [reply src/modules/ztsi/tests/ZtsiTests.cpp:91]: Same for next

**Comment by MariusNi** [comment]
> static const std::string g_urlRegex = "((http|https)://)(www.)?[-A-Za-z0-9+&@#/%?=~_|!:,.;]+[-A-Za-z0-9+&@#/%=~_|]";

This is only used in Ztsi::IsValidConfiguration. Make it local there? Same may apply to other global constants here.

---
Refers to: src/modules/ztsi/src/lib/Ztsi.cpp:27 in ce3be78. [](commit_id = ce3be78f1b17d41fafe87f69022f87bf09975758, deletion_comment = False)

File: src/modules/ztsi/tests/ZtsiTests.cpp (hunk 4 above)
Comment (robertschaedler3) [reply src/modules/ztsi/tests/ZtsiTests.cpp:91]: Added here and to Sample CPP module

**Review by MariusNi — APPROVED** [review]
:shipit:
